### PR TITLE
Nosqlbench 1522 moremath

### LINKED ISSUE #1522: support additional relevancy metrics in computation module

### PR DESCRIPTION
**Pull request by jshook**
fixes #1522

### CODE REVIEW

**Review by MarkWolters — APPROVED** [review]

File: engine-extensions/src/main/java/io/nosqlbench/engine/extensions/computefunctions/ComputeFunctions.java [hunk 1]
```diff
@@ -146,14 +156,66 @@ public static int[] intersection(int[] reference, int[] sample, int limit) {
...
         return Intersections.count(reference, sample, limit);
     }
 
+    public static double F1(int[] reference, int[] sample) {
+        return F1(reference, sample, reference.length);
+    }
+
+    public static double F1(int[] reference, int[] sample, int limit) {
+        double recallAtK = recall(reference, sample, limit);
+        double precisionAtK = precision(reference, sample, limit);
```
Comment (dave2wave) [inline engine-extensions/src/main/java/io/nosqlbench/engine/extensions/computefunctions/ComputeFunctions.java:178]: Is there a way to reuse prior metric results from the same observation? Or does that introduce complexity that would eliminate any perceived performance gain?

File: engine-extensions/src/main/java/io/nosqlbench/engine/extensions/computefunctions/Intersections.java [hunk 2]
```diff
@@ -20,6 +20,32 @@
 
 public class Intersections {
 
+    /**
+     * Return a non-negative index of the first value in the sample array which is present in the reference array,
+     * OR, return a negative number. This returns array index which start at 0, not rank, which is starts at 1.
+     */
+    public static int firstMatchingIndex(long[] reference, long[] sample, int limit) {
+        Arrays.sort(reference);
```
Comment (dave2wave) [inline engine-extensions/src/main/java/io/nosqlbench/engine/extensions/computefunctions/Intersections.java:28]: If this is almost always near the beginning and our limits are generally small then a brute force check of the reference list without sorting and binarysearch may be more efficient.

**Review by dave2wave — APPROVED** [review]
LGTM, but the findFirst could be better without sorting and searching and just iterating - especially if the number is almost always 1.

File: engine-extensions/src/main/java/io/nosqlbench/engine/extensions/computefunctions/Intersections.java (hunk 2 above)
Comment (jshook) [reply engine-extensions/src/main/java/io/nosqlbench/engine/extensions/computefunctions/Intersections.java:28]: Good point. We do expect it to be biased towards lower indices. Interesting optimization could happen here, but these array ops are already pretty freakin fast. Let's see if it catches on fire first.

**Pull request merged**